### ISSUE TITLE
Add `exists` Remap language function to check if event field exists

### ISSUE THREAD

**Issue by JeanMertz** [timberio/vector#4947]
Right now, if you reference a field that doesn't exist, a `null` value is returned:

```json
{ "foo": "bar" }
```

```rust
.baz // ok: null
```

While this is convenient (the alternative would be to abort the program with an error, which is too invasive), it does make it impossible to know if a field is present in an event by checking the result of the path query:

```json
{ "foo": null }
```

```rust
.foo // ok: null
.baz // ok: null
```

This issue proposes adding an `exists` function:

```rust
fn exists(Path) -> bool
```

This function takes a single path expression parameter, and returns a boolean value, indicating if the path exists or not.

Some examples:

```json
{
	"a": null,
	"b": [0, 1, 2],
	"c": {
		"d": 10
	},
	"e": [
		{ "f": false }
	]
}
```

```rust
exists(.a)		// ok: true
exists(.b)		// ok: true
exists(.c)		// ok: true
exists(.c.d)	// ok: true
exists(.e)		// ok: true
exists(.e[0])	// ok: true
exists(.e[0].f)	// ok: true

exists(.z)		// ok: false
exists(.c.z)	// ok: false
exists(.e[1])	// ok: false

exists(".a")	// compile error: required path argument, got literal
```

This then allows distinguishing between `null` fields and non-existing fields in programs:

```rust
if .a {							// failed condition, because value is `null`, which is treated as `false`
	"foo"
} else if exists(.a) {		// passed condition, because `.a` exists, function returns `true`
	"bar"
}
```

cc @binarylogic @FungusHumungus

**Comment by binarylogic** [comment]
Missing paths should not error, they should return `null`. I'd support an `exists` function to detect if a field exists or not though.

**Comment by JeanMertz** [comment]
@binarylogic I agree on not erroring.

On returning `null`, that's doable (but might be surprising).

If we were to add an `exists` function, and paths return `null` if they don't exist, you'd be required to write `exists(".foo")` (note the string, instead of `exists(.foo)`). This is because otherwise `.foo` is resolved before its value (`null`) is passed to the `exists` function.

**Comment by binarylogic** [comment]
Yeah, that's confusing and we'd definitely need to support the `exists(.foo)` syntax somehow (without the quotes).

**Comment by JeanMertz** [comment]
@binarylogic note that `del` and `only_fields` currently also require the quotes:

```rust
del(".foo", ".bar")
only_fields(".baz")
```

We could probably add a new type of function trait, which only accepts paths. It would be special-cased in the parser, meaning it wouldn't treat `.foo` as a path expression that needs to be resolved, but instead passes on the actual path as a string to the function.

So, functions such as `del`, `merge`, `only_fields` and `exists` would error if you provide anything other than a path, and they would have a custom implementation that differs from all the other "regular" functions that take expressions as their arguments.

**Comment by binarylogic** [comment]
Ah, I was not aware of that. Yes, I find that very confusing and will definitely be a source of confusion for users. The original spec used paths, I believe. I'd prefer it if we could special-case these and find a way to support undefined paths.

**Comment by JeanMertz** [comment]
I would note that it does make the language more complex to reason about though, without having any knowledge about these functions.

That is to say:

```rust
.foo = "foobar"
contains(.foo, "bar")
exists(.foo)
```

For `contains`, it would actually evaluate `.foo` to `"foobar"`, whereas `exists` wouldn't do anything with the `"foobar"` value.

I'm fine doing this, but I want to make sure you're aware of the complexity that can create when reading a remap script.

**Comment by binarylogic** [comment]
I'm not sure I understand. The example above seems straightforward to me.

```
.foo = "foobar"
contains(.foo, "bar") -> false
exists(.foo) -> true
```

**Comment by FungusHumungus** [comment]
I think the confusion could come because `contains` can take a string as well as a path, but `exists` can only take a path.

So  `contains("barbar", "bar")` is valid, but `exists("bar")` is not.

**Comment by JeanMertz** [comment]
The confusion can come from the fact that one function evaluates `.foo`, the other doesn't. I agree that this isn't the best example, and perhaps whatever confusing example we can come up with isn't as common that it doesn't matter, but I wanted to make sure we're aware that there are fundamental differences in how these two functions treat their arguments.

Here's one example that _might_ be confusing:

```rust
.foo = "foobar"
contains(.foo, "bar") -> true
exists(.foo) -> true

$foo = "foobar"
contains($foo, "bar") -> true
exists($foo) -> compiler error: expected path

contains("foobar", "bar") -> true
exists("foobar") -> compiler error: expected path

```

_(I added the third example, because we might want to actually support paths or variables in these special functions)_

I think I'm actually fine with this, since the error happens at compile-time, but I want to make sure we're aware of the implications, and we'll also need to make sure we differentiate between these "path-arguments only" functions and "expression argument" functions in the documentation.

I guess what I'm saying is this: http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/

But instead, in our case we don't have async and non-async functions, but we have "path-only" and "expression" functions.

**Comment by JeanMertz** [comment]
Technically, we _can_ support `exists(.foo)` using the existing infrastructure.

In the `exists` (and other) function, we could downcast the expression to check if we're dealing with a `Path` expression. If we are, we don't evaluate the expression, but instead use the path identifier to see if the path exists in the event (or if the variable exists, if `exists` is called with a variable expression).

This has upsides (we don't need to special case this in the language itself, `exists` is just another function), but it also has downsides, as the original confusion over which function resolves paths and which doesn't still exists.

**Comment by binarylogic** [comment]
For what it's worth. I never realized paths were evaluated before being passed down to functions, I assumed functions would handle paths as they see fit. That's how it was presented in the [original spec](https://github.com/timberio/vector/blob/master/rfcs/2020-07-21-2744-remap-syntax.md#guide-level-proposal).

**Comment by JeanMertz** [comment]
> I assumed functions would handle paths as they see fit

Your assumption is correct. That _is_ the case. But Paths are passed to functions as _expressions_, similar to how literals, variables, and other functions are passed in as expressions for function arguments.

Until now, all functions simply resolve those expressions to their final value, as that's what we needed so far.

When we moved the hard-coded `del` and `only_fields` functions from the parser to regular types that implement the `Expression` trait, we switched from `del(.foo)` to `del(".foo")` to avoid the confusion we discussed above.

However, other than the possible confusion, there's no technical reason we can't treat paths as-is within a function, which is also why I proposed we _can_ do it with the existing infrastructure, by downcasting the expression back to a path.

The one downside to this (right now) is that passing a non-path expression is a runtime error instead of a compile-time error.

With the work we're doing to move more error checking to the compile-step, it can eventually become a compile-time error again though.

The alternative solution I proposed is hard-coding these functions in the parser, which gives us compile-time errors, but also makes things more complex/complicated on a technical level, since we now have to distinguish between "regular functions" and "hard-coded functions" again.

I would say we start with `exists(.foo)` that returns a runtime error if you try to pass in `exists("foo")`. Once the compile-check infrastructure is further along, we can have it make sure you pass in a `Path` expression, and nothing else at compile-time.

**Comment by JeanMertz** [comment]
I updated this issue to more accurately describe our current thinking.

Things that have changed since I created this issue:

* Querying fields that don't exist return a `null` value. #5056
* We now support requiring _path expressions_ as function parameters (e.g. `del(.foo)`) #5015

Given this. The next step would be to add an `exists` function, which takes a single path, and returns a boolean value, indicating whether the field is present in the event.

**Comment by JeanMertz** [comment]
One thing I would add, is that down the line, if we feel like calling `exists` is something many users of Remap do, we can introduce a special sigil to make this more convenient to check.

For example, I can see something like this working well:

```rust
exists(.foo.bar.baz) 	// ok: false
.foo.bar.baz 			// ok: null
.boo.bar.baz?			// ok: false
```

That is to say, by appending the `?` question-mark sigil at the end of a path query, the query no longer returns the value of the path, but a boolean indicating if the path exists or not.

If one has a field named `baz?`, then the quoted path variant can be used to distinguish between the two:

```rust
.foo.bar.baz? 	// ok: false
.foo.bar."baz?" // ok: null
```

_(the requirement to use quoted fields is already in effect today, since we don't support any special characters in regular paths, precisely to give us more flexibility to use those characters for special cases such as this one)_

We don't need to do this in the first iteration, but checking for the existence of fields seems like a fairly ubiquitous task that one might want to do as convenient as possible.

**Comment by lukesteensen** [comment]
> The next step would be to add an `exists` function, which takes a single path, and returns a boolean value, indicating whether the field is present in the event.

Before we commit to this, I think it's worth considering whether we could make it more generic and operate not just on paths but the return values of methods like `parse_json`. They're similar in that the result could be "nothing"/none/undefined, and it would be nice to have unified handling of them if we can come up with something natural.

---

> I can think of two downsides to this secondary proposal

I would add a third in that it's relatively unexpected given the way that sigil is used in other languages. Rust's error handling is the most obvious example for us, but it also looks very much like the [safe navigation operator](https://en.wikipedia.org/wiki/Safe_navigation_operator) present in many languages.

The primary benefit of using a sigil over something like a function is the ability to chain calls, which isn't really something you need to do with a boolean.

**Comment by JeanMertz** [comment]
> They're similar in that the result could be "nothing"/none/undefined

The problem is that you want to be able to distinguish between null and non-existing. That's mostly what this issue is about.

I presume by specifically writing none/undefined, you're actually talking about "not having any value at all". It is true that we did support this before, but we backed out of that (#5053) after some offline discussions, since it made other parts of the language harder to reason about.

Specifically, some functions considered `none` as input to be an error, while others passed it along without processing the input. Aside from functions, it also was unclear what each built-in operation should do with none, such as assignments receiving none, none in if statements, etc.

_If_ we were to reintroduce the concept of "undefined", I would definitely propose we RFC it first, so that we consider all downsides along the upsides.

> The primary benefit of using a sigil over something like a function is the ability to chain calls

I would note that the primary benefit I listed is the brevity it introduces.

I've written a medium-sized LUA transform before, and had many `if .foo.bar.baz` checks in them to handle events from different sources in a uniform way.

I'd rather do `if .foo.bar?` a bunch of times vs. `if exists(.foo.bar)`.

Having said that, I agree that it's not all roses, and as mentioned, I also don't think we should tackle this problem in the first iteration. Especially, as you mentioned, since we might be able to align this with error handling in some way.
